prog/dftb+/lib_dftbplus/parser.F90

```diff
@@ -3516,6 +3536,139 @@ subroutine readOptions(node, ctrl)
+    type(TGeometry), intent(in) :: geo
+
+    !> Contains the input for the dispersion module on exit
+    type(TGBInput), intent(out) :: input
+
+    type(string) :: buffer, state, modifier
+    type(fnode), pointer :: child, value1, field, child2
+    integer :: iSp
+    logical :: found
+    real(dp) :: temperature, shift
+    real(dp), allocatable :: vdwRadDefault(:)
+    type(TSolventData) :: solvent
+
+    if (geo%tPeriodic) then
+      call detailedError(node, "Generalized Born model currently not available with PBCs")
+    end if
+
+    call getChildValue(node, "Solvent", value1, child=child)
+    call getNodeName(value1, buffer)
+    select case(char(buffer))
+    case default
+      call detailedError(child, "Invalid solvent method '" // char(buffer) // "'")
+    case('fromname')
+      call getChildValue(value1, "", buffer)
+      call SolventFromName(solvent, unquote(char(buffer)), found)
+      if (.not. found) then
+        call detailedError(value1, "Invalid solvent " // char(buffer))
+      end if
+    case('fromconstants')
+      call getChildValue(value1, "Epsilon", solvent%dielectricConstant)
+      call getChildValue(value1, "MolecularMass", solvent%molecularMass)
+      call getChildValue(value1, "Density", solvent%density)
+    end select
+
+    input%keps = 1.0_dp / solvent%dielectricConstant - 1.0_dp
+
+    ! shift value for the free energy (usually fitted)
+    call getChildValue(node, "Shift", shift, modifier=modifier, child=field)
+    call convertByMul(char(modifier), energyUnits, field, shift)
+
+    ! temperature, influence depends on the reference state
+    call getChildValue(node, "Temperature", temperature, 298.15_dp)
+
+    ! reference state for free energy calculation
+    call getChildValue(node, "State", state, "gsolv", child=child)
+    select case(tolower(unquote(char(state))))
+    case default
+      call detailedError(child, "Unknown reference state: '"//char(state)//"'")
+    case("gsolv") ! just the bare shift
+      input%shift = shift
+    end select
+
+    call getChildValue(node, "BornScale", input%bornScale)
+    call getChildValue(node, "BornOffset", input%bornOffset, modifier=modifier, child=field)
+    call convertByMul(char(modifier), lengthUnits, field, shift)
+    call getChildValue(node, "OBC", input%obc, [1.00_dp, 0.80_dp, 4.85_dp])
+
+    allocate(input%vdwRad(geo%nSpecies))
+    call getChildValue(node, "Radii", value1, child=child)
+    call getNodeName(value1, buffer)
+    select case(char(buffer))
+    case default
+      call detailedError(child, "Unknown method '"//char(buffer)//"' to generate radii")
+    case("vanderwaalsradiid3")
+      allocate(vdwRadDefault(geo%nSpecies))
+      vdwRadDefault(:) = getVanDerWaalsRadiusD3(geo%speciesNames)
+      do iSp = 1, geo%nSpecies
+        call getChildValue(value1, geo%speciesNames(iSp), input%vdwRad(iSp), &
+            & vdwRadDefault(iSp), child=child2)
+      end do
+      deallocate(vdwRadDefault)
+    case("values")
```

Comment:
Back to debate about units in these cases, probably should carry one here.

Reply:
I've a proposal for modifiers here: d96d163. It's a bit hacky, but works in a robust way.

Reply:
Having the modifier at `vanDerWaalsRadiiD3` or `values` instead of `radii` would be preferable, but I don't see how to do this with the `getChildValue` routines.

Reply:
See 569fef6. :wink: 

Unfortunately, playing with it drew my attention to a bug in our current parser API: The parser allowed for a modifier on `VanDerWaalsRadiiD3` in the previous implementation, it was just ignored. For the current case, it is fixed now, but way have several places in the input, where an accidently specified modifier may be silently ignored, instead of triggering an error.....

Reply:
So every construct like
```cson
key {
  method {
    A = ...
    B = ...
    C = ...
  }
}
```
can carry an untracked modifier at the `method` level?